### PR TITLE
Strict mode

### PR DESCRIPTION
**Pull request by kurobako**
-Dorg.kie.dmn.strictConformance or -Dorg.kie.dmn.strictConformance=true
-Ptck always runs in strict mode

### CODE REVIEW

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELDateTimeDurationTest.java [hunk 1]
```diff
@@ -163,15 +163,7 @@
                 { "years and months duration( date(\"2011-12-22\"), date(\"2013-08-24\") ).years", BigDecimal.valueOf(1) , null},
                 { "years and months duration( date(\"2011-12-22\"), date(\"2013-08-24\") ).months", BigDecimal.valueOf(8) , null},
                 { "date and time(\"2017-05-14\")", LocalDateTime.of( 2017, 5, 14, 0, 0, 0, 0 ) , null},
-                { "date(\"2017-05-12\")-date(\"2017-04-25\")", Duration.ofDays( 17 ) , null},
-
-                // the following is an extension to the standard: DROOLS-1549
-                { "date(\"2016-12-20T14:30:22\")", DateTimeFormatter.ISO_DATE.parse( "2016-12-20", LocalDate::from ) , null },
```
Comment (baldimir) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELDateTimeDurationTest.java]: These should be tested somewhere, so we are testing also our extensions to the standard. That applies also to other removed test cases from other classes.

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/extended/DateFunction.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                .appendLiteral('-')
+                .appendValue(DAY_OF_MONTH, 2)
+                .toFormatter();
+    }
+
+    DateFunction() {
+        super("date");
+    }
+
+    public FEELFnResult<TemporalAccessor> invoke(@ParameterName("from") String val) {
```
Comment (baldimir) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/extended/DateFunction.java:38]: These new function classes are copies of the originals we had. This creates duplicate code that could be harder to maintain in the future. E.g. when the definition of a function changes slightly, we might forget that we need to change that at 2 places. Would it be possible to remove the duplication somehow? E.g. by creating abstract function classes that will contain the mutual code and then inheriting strict and extended versions?

Comment (kurobako) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/extended/DateFunction.java:38]: This is a conscious choice. OOP is close to useless with regards to code deduplication in such scenario since extra complexity created by 'ExtensibleDateFunction' with a bunch of methods to override would imo overweight the value it brings.

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELDateTimeDurationTest.java (hunk 1 above)
Comment (kurobako) [reply kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELDateTimeDurationTest.java]: Take a look at ExtendedDMNProfileTest.java
If that's the wrong=unexpected place for such tests in your opinion that's fine, I can move those to wherever you want them tests to be, just let me know.

Comment (baldimir) [reply kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELDateTimeDurationTest.java]: Ok sorry, I'm blind. Thanks for clarification.

**Comment by kurobako** [comment]
@baldimir updated my PR, including your content
mvn clean install -Dtck.repository.url=https://github.com/actico/tck.git -Dtck.repository.branch=generated-testcases-2017-12-12 -Ptck -Dmaven.test.skip runs fine now
tck master fails since it has opposite idea about some tests

can you please try applying patch of this PR and running mvn clean install locally if you don't mind?

**Comment by baldimir** [comment]
@kurobako will check your changes. I am also trying one approach to remove the duplicate code. Will let you know in the afternoon.

File: kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/compiler/extensions/DMNExtensionRegisterTest.java [hunk 3]
```diff
@@ -46,6 +47,7 @@
 public class DMNExtensionRegisterTest {
     private static final Logger LOG = LoggerFactory.getLogger(DMNExtensionRegisterTest.class);
 
+    @Ignore
```
Comment (baldimir) [inline kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/compiler/extensions/DMNExtensionRegisterTest.java:50]: Just curious, why are these tests Ignored?

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/FunctionTestUtil.java [hunk 4]
```diff
@@ -31,7 +31,7 @@
             assertResultNotError(result);
             final T resultValue = result.cata(left -> null, right -> right);
             Assert.assertThat(resultValue, Matchers.notNullValue());
-            Assert.assertThat(resultValue, Matchers.equalTo(expectedResult));
```
Comment (baldimir) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/FunctionTestUtil.java]: Why is this removed?

**Comment by baldimir** [comment]
@kurobako I ran your changes locally, I see 2 test failures and some failures in kie-dmn-validation tests. I have fixes for the 2 test failures and will take a look at the validation module. Also I have removed the code duplication. Will link my changes when ready (should be in an hour or so, after I fix validation module). We can discuss the duplication removal then.

File: kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/compiler/extensions/DMNExtensionRegisterTest.java (hunk 3 above)
Comment (kurobako) [reply kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/compiler/extensions/DMNExtensionRegisterTest.java:50]: @baldimir we had a discussion with Matteo and decided ultimately that he will take a look at these (enabling them back I guess) after my PR will be done. It's either tests have to be redone or server behaviour has to be changed when profiles (extended profile in our case) are on, but has not much to do with what this PR is about, hence the delay.

**Comment by baldimir** [comment]
@kurobako here are my proposed changes [1]. The main thing is having Base classes with all the logic and then the functions which just call the invokes from the Base classes. This could be further improved if the call that searches for invoke methods [2] will be changed to getMethods(). In such case, majority of invoke definitions could be removed from the inherited functions. However I am not sure about the performance impact of such change, that's why I left it as it is. getDeclaredMethods() searches just the top class for methods, getMethods() on the other hand searches the whole inheritance tree. In such case it would iterate over more methods, so the invocation could be slower. 

To fix the problems in module kie-dmn-validation I had to rebase the branch to latest master. I'm sorry but because of that I cannot show just my changes, I had to link comparison with Drools master. 

The other 2 test failures were just test leftovers or typos. However please, if we decide at the end that we want to have the duplicated code, add at least these fixes.

@tarilabs @etirelli please, also take a look.  

[1] https://github.com/kiegroup/drools/compare/master...baldimir:StrictMode?expand=1
[2] https://github.com/baldimir/drools/blob/8c5a8c8084a0bda650977fe37bc5630261ea8a6e/kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/BaseFEELFunction.java#L190

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/FunctionTestUtil.java (hunk 4 above)
Comment (kurobako) [reply kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/FunctionTestUtil.java]: @baldimir wtf I am not sure why it is even here. Thanks for noticing :+1:  I will update my PR, rolling this change back in a second.

**Comment by baldimir** [comment]
I also tested TCK with changes from @kurobako + my changes [1] and all tests pass. 

[1] https://github.com/kiegroup/drools/compare/master...baldimir:StrictMode?expand=1

**Comment by tarilabs** [comment]
I see @baldimir point, and thank you for the [1] reference.
But I must admit I am conflicted.
On one side this indeed avoids code "duplication", between the Base which should be 100% DMN conformance level 3 adherent, and our "drools, non-strict" version of the same functions.
On other side however this is actually making the "drools, non-strict" code version still in the base-code, enabled with the boolean; like saying our own extensions to the standard have a different treatment if compared to a common extension.

@kurobako maybe I missed it, but is there a test showing using the flag as a kmodule.xml property now, please?

**Comment by kurobako** [comment]
@baldimir updated my PR again, including your fixes (thanks!). Both generated-testcases-2017-12-12 tck and unit tests work fine for me.
Regarding your deduplication suggestion, I don't mind. I mean, I am still not convinced it's the way to go but anyway I think it makes sense to merge my PR first and then you can make yours (and with a proper authorship).

@tarilabs I was quite busy with this ever changing state of TCK I don't even want to comment on and some unit test failures so I didn't work on this yet. I think this can be made as a next PR. I don't think we should delay this one because of kmodule. Wdyt?

**Comment by baldimir** [comment]
@kurobako thanks for updates. I am also fine both ways. As @tarilabs mentioned both things have pros and cons. It is just a suggestion. So +1 from me if we decide to merge it as it is.

**Comment by kurobako** [comment]
13:21:42 Unexpected error while executing the Upstream repositories builder for PR builds! null
13:21:42 java.lang.NullPointerException
13:21:42 Build step 'Build required upstream repositories (for PR builds)' marked build as failure

nevermind the CI

**Comment by tarilabs** [comment]
Jenkins test again

**Comment by tarilabs** [comment]
@kurobako where you say:
> @tarilabs I was quite busy with this ever changing state of TCK I don't even want to comment on and some unit test failures so I didn't work on this yet. I think this can be made as a next PR. I don't think we should delay this one because of kmodule. Wdyt?

I was asking as there are no unit test validating the disable/enable of this flag.
Besides leveraging the TCK module which the "side effect" of passing the flag as a sys property exhibit as behavior, but in order words is not unit tested.
Leaving kmodule a side for the moment, if a unit test could show this using a system property, I thought would have been nice for validation of this change (beside leveraging the TCK).

I guess is now up to @etirelli for considering the point of @baldimir above, and if the PR is anyway better to be merged early. And add unit test of this flag, later.

**Comment by kurobako** [comment]
@tarilabs -Ptck does exactly this: it runs jvm with the flag.
I will write unit test a bit later, yes.

**Comment by baldimir** [comment]
jenkins retest this

**Comment by baldimir** [comment]
jenkins retest this

**Comment by kurobako** [comment]
@tarilabs I updated this PR with unit test as you suggested that checks system property.
Regarding kmodules: I took a look at what now ignored DMNExtensionRegisterTes does with kmodules and from what I see they are broken. I suggest fixing them first in a separate PR.Kmodule xml is written to to one instance of filesystem and then builder tries to read xml from another instance of filesystem that knows nothing of it and therefore creates new empty module.
I will continue investigating next week if @etirelli won't mind, but absolutely not in this PR. I consider this one done.

**Comment by baldimir** [comment]
jenkins retest this

**Comment by baldimir** [comment]
@kurobako I tried to compile your branch again locally and it doesn't compile. There were probably some changes in related repositories. I tried to rebase it then and it compiles with the latest changes from master. So please, rebase your branch again. Maybe that is why the PR builder fails. If it not helps, I will ask our Jenkins guys.

**Comment by kurobako** [comment]
@baldimir that's odd. But anyway, I just rebased my PR against master once again. Let's see if CI will be kind to us this time.

**Comment by baldimir** [comment]
jenkins retest this

**Comment by kurobako** [comment]
13:23:04 Unexpected error while executing the Upstream repositories builder for PR builds! null
13:23:04 java.lang.NullPointerException
13:23:04 Build step 'Build required upstream repositories (for PR builds)' marked build as failure

**Comment by baldimir** [comment]
jenkins retest this

**Comment by baldimir** [comment]
jenkins retest this

**Comment by tarilabs** [comment]
Couldn't it be the CI issues are related to the fact that this PR is originating from a branch name called `master` ?
( from kurobako:master ) 
I don't know if this potentially influence how Jenkins is supposed to manage it ?

**Comment by kurobako** [comment]
updated my PR so now it's possible to configure strict mode using kmodule props. Unit test added as well.

@etirelli is it good for merge now?

**Comment by kurobako** [comment]
[INFO] Reactor Summary:
[INFO] 
[INFO] Drools Multiproject ................................ SUCCESS [  2.344 s]
[INFO] KIE :: Test Utility Classes ........................ SUCCESS [  5.194 s]
[INFO] Drools :: Core ..................................... SUCCESS [01:03 min]
[INFO] Drools :: Compiler ................................. SUCCESS [03:58 min]
[INFO] Drools :: Beliefs .................................. SUCCESS [  4.192 s]
[INFO] Drools :: CDI ...................................... SUCCESS [ 12.728 s]
[INFO] Drools :: Verifier ................................. SUCCESS [ 33.743 s]
[INFO] Drools Persistence projects ........................ SUCCESS [  0.052 s]
[INFO] Drools :: Persistence :: API ....................... SUCCESS [  0.508 s]
[INFO] Drools :: Persistence :: JPA ....................... SUCCESS [ 24.316 s]
[INFO] Drools :: Templates ................................ SUCCESS [  2.229 s]
[INFO] Drools :: Decision Tables .......................... SUCCESS [ 13.470 s]
[INFO] Drools examples .................................... SUCCESS [  1.428 s]
[INFO] Drools :: PMML - Compiler .......................... SUCCESS [01:24 min]
[INFO] Drools :: Scorecards ............................... SUCCESS [ 28.123 s]
[INFO] KIE :: CI .......................................... SUCCESS [02:17 min]
[INFO] KIE :: CI for OSGi ................................. SUCCESS [  2.296 s]
[INFO] KIE :: Decision Model Notation ..................... SUCCESS [  0.205 s]
[INFO] KIE :: Decision Model Notation :: Model ............ SUCCESS [  0.402 s]
[INFO] KIE :: Decision Model Notation :: API .............. SUCCESS [  0.323 s]
[INFO] KIE :: Decision Model Notation :: FEEL ............. SUCCESS [ 11.069 s]
[INFO] KIE :: Decision Model Notation :: Backend .......... SUCCESS [  2.777 s]
[INFO] KIE :: Decision Model Notation :: Core ............. SUCCESS [ 12.352 s]
[INFO] KIE :: Decision Model Notation :: Validation ....... SUCCESS [ 22.222 s]
[INFO] KIE :: Decision Model Notation :: TCK .............. SUCCESS [  0.043 s]
[INFO] KIE :: Decision Model Notation :: Signavio extensions support SUCCESS [  2.454 s]
[INFO] KIE :: PMML - Compiler ............................. SUCCESS [ 52.591 s]
[INFO] Drools API examples ................................ SUCCESS [  0.042 s]
[INFO] Drools API examples - Default KieSession ........... SUCCESS [  1.571 s]
[INFO] Drools API examples - Named KieSession ............. SUCCESS [  1.874 s]
[INFO] Drools API examples - KieBase Inclusion ............ SUCCESS [  1.437 s]
[INFO] Drools API examples - Default KieSession From File . SUCCESS [  1.467 s]
[INFO] Drools API examples - Named KieSession from File ... SUCCESS [  1.418 s]
[INFO] Drools API examples - KieModule from Multiple Files  SUCCESS [  1.524 s]
[INFO] Drools API examples - KieFileSystemExample ......... SUCCESS [  1.430 s]
[INFO] Drools API examples - KieModuleModel Example ....... SUCCESS [  1.882 s]
[INFO] Drools API examples - KieContainer from KieRepo .... SUCCESS [  5.780 s]
[INFO] Drools API examples - Multiple KieBases ............ SUCCESS [  1.601 s]
[INFO] Drools API examples - Reactive KieSession .......... SUCCESS [  1.736 s]
[INFO] Drools API examples - RuleUnit ..................... SUCCESS [  1.547 s]
[INFO] Drools CDI examples ................................ SUCCESS [  0.037 s]
[INFO] Drools CDI Example ................................. SUCCESS [  2.140 s]
[INFO] Drools CDI Example With Inclusion .................. SUCCESS [  2.034 s]
[INFO] cdi-example-scopes ................................. SUCCESS [  3.924 s]
[INFO] Drools Workbench - Models .......................... SUCCESS [  0.035 s]
[INFO] Drools Workbench - Data Model API .................. SUCCESS [  5.298 s]
[INFO] Drools Workbench - Common Model .................... SUCCESS [  2.557 s]
[INFO] Drools Workbench - Guided Decision Table Model ..... SUCCESS [  3.722 s]
[INFO] Drools Workbench - Guided Decision Tree Model ...... SUCCESS [  1.508 s]
[INFO] Drools Workbench - Guided Rule Templates Model ..... SUCCESS [  5.997 s]
[INFO] Drools Workbench - Guided Score Cards Model ........ SUCCESS [ 11.186 s]
[INFO] Drools Workbench - Test Scenarios Model ............ SUCCESS [  3.167 s]
[INFO] drools-model ....................................... SUCCESS [  0.043 s]
[INFO] drools-canonical-model ............................. SUCCESS [  1.180 s]
[INFO] drools-model-compiler .............................. SUCCESS [ 32.202 s]
[INFO] Drools :: Test Coverage ............................ SUCCESS [  0.039 s]
[INFO] Drools :: Test Coverage :: Standalone reproducers .. SUCCESS [  0.037 s]
[INFO] Drools :: Test Coverage :: KIE-CI with Domain classes :: Parent SUCCESS [  0.038 s]
[INFO] Drools :: Test Coverage :: KIE-CI with Domain classes :: Domain SUCCESS [  0.120 s]
[INFO] Drools :: Test Coverage :: KIE-CI with Domain classes :: KJar SUCCESS [  0.100 s]
[INFO] Drools :: Test Coverage :: KIE-CI with Domain classes :: Tests SUCCESS [  6.476 s]
[INFO] Drools :: Test Coverage :: KIE-CI without Domain classes :: Parent SUCCESS [  0.137 s]
[INFO] Drools :: Test Coverage :: KIE-CI without Domain classes :: Domain SUCCESS [  0.176 s]
[INFO] Drools :: Test Coverage :: KIE-CI without Domain classes :: KJar SUCCESS [  0.185 s]
[INFO] Drools :: Test Coverage :: KIE-CI without Domain classes :: Tests SUCCESS [  7.743 s]
[INFO] Drools :: Test Coverage :: Test Suite .............. SUCCESS [02:12 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 16:52 min
[INFO] Finished at: 2018-01-22T15:08:59+01:00
[INFO] Final Memory: 372M/1808M
[INFO] ------------------------------------------------------------------------

**Comment by tarilabs** [comment]
Concerning the CI failing, I reiterate my suggestion from last week. I can't be certain that is the problem, but imho is worth a try.

**Comment by kurobako** [comment]
new PR: https://github.com/kiegroup/drools/pull/1711
